### PR TITLE
[Merged by Bors] - feat(Combinatorics/SimpleGraph): add miscellaneous lemmas

### PR DESCRIPTION
**Pull request by Command-Master**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Dart.lean [hunk 1]
```diff
@@ -39,6 +39,10 @@ theorem Dart.ext (d₁ d₂ : G.Dart) (h : d₁.toProd = d₂.toProd) : d₁ = d
 -- Porting note: deleted `Dart.fst` and `Dart.snd` since they are now invalid declaration names,
 -- even though there is not actually a `SimpleGraph.Dart.fst` or `SimpleGraph.Dart.snd`.
 
+@[simp]
+theorem Dart.ne (d : G.Dart) : d.toProd.1 ≠ d.toProd.2 :=
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Dart.lean:43]:
```suggestion
theorem Dart.fst_ne_snd (d : G.Dart) : d.fst ≠ d.snd :=
```
and you might also add `snd_ne_fst`

File: Mathlib/Combinatorics/SimpleGraph/Walk.lean [hunk 2]
```diff
@@ -494,6 +495,17 @@ theorem support_reverse {u v : V} (p : G.Walk u v) : p.reverse.support = p.suppo
 @[simp]
 theorem support_ne_nil {u v : V} (p : G.Walk u v) : p.support ≠ [] := by cases p <;> simp
 
+@[simp]
+theorem support_head {G : SimpleGraph V} {a b : V} (p : G.Walk a b) :
+    p.support.head (by simp) = a := by cases p <;> simp
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Walk.lean:500]:
```suggestion
theorem head_support {G : SimpleGraph V} {a b : V} (p : G.Walk a b) :
    p.support.head (by simp) = a := by cases p <;> simp
```

File: Mathlib/Combinatorics/SimpleGraph/Walk.lean [hunk 3]
```diff
@@ -494,6 +495,17 @@ theorem support_reverse {u v : V} (p : G.Walk u v) : p.reverse.support = p.suppo
...
+
+@[simp]
+theorem support_getLast {G : SimpleGraph V} {a b : V} (p : G.Walk a b) :
+    p.support.getLast (by simp) = b := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Walk.lean:504]:
```suggestion
theorem getLast_support {G : SimpleGraph V} {a b : V} (p : G.Walk a b) :
    p.support.getLast (by simp) = b := by
```

File: Mathlib/Combinatorics/SimpleGraph/Walk.lean [hunk 4]
```diff
@@ -636,6 +648,21 @@ theorem map_fst_darts_append {u v : V} (p : G.Walk u v) :
 theorem map_fst_darts {u v : V} (p : G.Walk u v) : p.darts.map (·.fst) = p.support.dropLast := by
   simpa! using congr_arg List.dropLast (map_fst_darts_append p)
 
+@[simp]
+theorem darts_head_fst {G : SimpleGraph V} {a b : V} (p : G.Walk a b) (hp : p.darts ≠ []) :
+    (p.darts.head hp).toProd.1 = a := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Walk.lean:653]:
```suggestion
theorem head_darts_fst {G : SimpleGraph V} {a b : V} (p : G.Walk a b) (hp : p.darts ≠ []) :
    (p.darts.head hp).fst = a := by
```
or
```suggestion
theorem fst_head_darts {G : SimpleGraph V} {a b : V} (p : G.Walk a b) (hp : p.darts ≠ []) :
    (p.darts.head hp).fst = a := by
```

**Comment by Command-Master** [comment]
I noticed that many of these lemmas are in #15711, so I think I'll mark it as a dependency

**Comment by YaelDillies** [comment]
Actually, #15711 is very big and unfocused, so I would rather mark it as depending on your PR, rather than the other way around.

**Comment by Command-Master** [comment]
Alright. @meow-sister, are you OK with that?

**Comment by meow-sister** [comment]
> Alright. @meow-sister, are you OK with that?

Sorry for reply lately. My PR would changes a lot in coding style and naming convention before merged.

**Comment by YaelDillies** [comment]
Does that mean you are happy with the current PR (#16294) being merged before yours (#15711)?

**Comment by meow-sister** [comment]
> Does that mean you are happy with the current PR (#16294) being merged before yours (#15711)?

In my opinion, the merging order is not so important.

**Comment by YaelDillies** [comment]
Okay. I have turned the dependency around

File: Mathlib/Combinatorics/SimpleGraph/Walk.lean [hunk 5]
```diff
@@ -641,6 +653,21 @@ theorem map_fst_darts_append {u v : V} (p : G.Walk u v) :
...
 
+@[simp]
+theorem head_darts_fst {G : SimpleGraph V} {a b : V} (p : G.Walk a b) (hp : p.darts ≠ []) :
+    (p.darts.head hp).fst = a := by
+  cases p
+  · contradiction
+  · simp
+
+@[simp]
+theorem darts_getLast_snd {G : SimpleGraph V} {a b : V} (p : G.Walk a b) (hp : p.darts ≠ []) :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Walk.lean:664]:
```suggestion
theorem getLast_darts_snd {G : SimpleGraph V} {a b : V} (p : G.Walk a b) (hp : p.darts ≠ []) :
```

**Comment by YaelDillies** [comment]
The rest looks good

maintainer merge

**Comment by bryangingechen** [comment]
This seems to have slipped off the queue
bors r+

**Pull request closed without merging**